source/fuzz/transformation_replace_opselect_with_conditional_branch.cpp

```diff
@@ -0,0 +1,146 @@
+
+  // Check that the new block ids are fresh and distinct.
+  std::set<uint32_t> used_ids;
+  for (uint32_t id :
+       {message_.new_block_ids().first(), message_.new_block_ids().second()}) {
+    if (!CheckIdIsFreshAndNotUsedByThisTransformation(id, ir_context,
+                                                      &used_ids)) {
+      return false;
+    }
+  }
+
+  // The block cannot be a loop header, since splitting it would make back edges
+  // invalid.
+  if (block->IsLoopHeader()) {
+    return false;
+  }
+
+  // The block must be split around the OpSelect instruction. This means that
+  // there cannot be an OpSampledImage instruction before OpSelect that is used
+  // after it, because they are required to be in the same basic block.
+  return !fuzzerutil::
+      SplitBeforeInstructionSeparatesOpSampledImageDefinitionFromUse(
+          block, instruction);
+}
+
+void TransformationReplaceOpSelectWithConditionalBranch::Apply(
+    opt::IRContext* ir_context, TransformationContext* /* unused */) const {
+  auto instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.select_id());
+
+  auto block = ir_context->get_instr_block(instruction);
+
+  // Update the module id bound with the new ids being used.
+  fuzzerutil::UpdateModuleIdBound(ir_context, message_.new_block_ids().first());
+  fuzzerutil::UpdateModuleIdBound(ir_context,
+                                  message_.new_block_ids().second());
+
+  // Split the block before the OpSelect instruction to get what will be the
+  // merge block.
+  auto merge_block = block->SplitBasicBlock(
+      ir_context, message_.new_block_ids().second(),
+      fuzzerutil::GetIteratorForInstruction(block, instruction));
+
+  // Create a new empty block.
+  auto new_block = MakeUnique<opt::BasicBlock>(MakeUnique<opt::Instruction>(
+      ir_context, SpvOpLabel, 0, message_.new_block_ids().first(),
+      opt::Instruction::OperandList{}));
+
+  // Add an unconditional branch from the new block to the merge block.
+  new_block->AddInstruction(MakeUnique<opt::Instruction>(
+      ir_context, SpvOpBranch, 0, 0,
+      opt::Instruction::OperandList{
+          {SPV_OPERAND_TYPE_ID, {merge_block->id()}}}));
+
+  // Add an OpSelectionMerge instruction to the original block.
+  block->AddInstruction(MakeUnique<opt::Instruction>(
+      ir_context, SpvOpSelectionMerge, 0, 0,
+      opt::Instruction::OperandList{
+          {SPV_OPERAND_TYPE_ID, {merge_block->id()}},
+          {SPV_OPERAND_TYPE_SELECTION_CONTROL, {0}}}));
+
+  // Add a conditional branching instruction to the original block, using the
+  // same conditional as OpSelect and branching to the new block if the
+  // condition is true, to the merge block otherwise.
+  block->AddInstruction(MakeUnique<opt::Instruction>(
+      ir_context, SpvOpBranchConditional, 0, 0,
+      opt::Instruction::OperandList{
+          {SPV_OPERAND_TYPE_ID, {instruction->GetSingleWordInOperand(0)}},
```

Comment:
Condition in OpSelect can be a scalar or a vector of boolean components. I think we should check the latter case in the IsApplicable since we don't support it.

Reply:
Yes, thanks for spotting this!